data-loader/core/src/main/java/com/scalar/db/dataloader/core/dataimport/controlfile/ControlFileValidator.java

```diff
@@ -0,0 +1,225 @@
+      }
+    }
+  }
+
+  /**
+   * Check that the required partition keys are mapped in the control file. Ran only for control
+   * file validation mode KEYS and FULL.
+   *
+   * @param tableMetadata Metadata for one ScalarDB table
+   * @param mappedTargetColumns Set of target columns that are mapped in the control file
+   * @param controlFileTable Control file entry for one ScalarDB table
+   * @throws ControlFileValidationException when a partition key is not mapped
+   */
+  private static void checkPartitionKeys(
+      TableMetadata tableMetadata,
+      Set<String> mappedTargetColumns,
+      ControlFileTable controlFileTable)
+      throws ControlFileValidationException {
+    LinkedHashSet<String> partitionKeyNames = tableMetadata.getPartitionKeyNames();
+    for (String partitionKeyName : partitionKeyNames) {
+      if (!mappedTargetColumns.contains(partitionKeyName)) {
+        throw new ControlFileValidationException(
+            CoreError.DATA_LOADER_MISSING_PARTITION_KEY.buildMessage(
+                partitionKeyName, TableMetadataUtil.getTableLookupKey(controlFileTable)));
+      }
+    }
+  }
+
+  /**
+   * Check that the required clustering keys are mapped in the control file. Ran only for control
+   * file validation mode KEYS and FULL.
+   *
+   * @param tableMetadata Metadata for one ScalarDB table
+   * @param mappedTargetColumns Set of target columns that are mapped in the control file
+   * @param controlFileTable Control file entry for one ScalarDB table
+   * @throws ControlFileValidationException when a clustering key is not mapped
+   */
+  private static void checkClusteringKeys(
+      TableMetadata tableMetadata,
+      Set<String> mappedTargetColumns,
+      ControlFileTable controlFileTable)
+      throws ControlFileValidationException {
+    LinkedHashSet<String> clusteringKeyNames = tableMetadata.getClusteringKeyNames();
+    for (String clusteringKeyName : clusteringKeyNames) {
+      if (!mappedTargetColumns.contains(clusteringKeyName)) {
+        throw new ControlFileValidationException(
+            CoreError.DATA_LOADER_MISSING_CLUSTERING_KEY.buildMessage(
+                clusteringKeyName, TableMetadataUtil.getTableLookupKey(controlFileTable)));
+      }
+    }
+  }
+
+  /**
+   * Check that a control file table mapping does not contain duplicate mappings for the same target
+   * column
+   *
+   * @param controlFileTable Control file entry for one ScalarDB table
+   * @return Set of uniquely mapped target columns
+   * @throws ControlFileValidationException when a duplicate mapping is found
+   */
+  private static Set<String> checkDuplicateColumnMappings(ControlFileTable controlFileTable)
+      throws ControlFileValidationException {
+    Set<String> mappedTargetColumns = new HashSet<>();
+    for (ControlFileTableFieldMapping mapping : controlFileTable.getMappings()) {
+      if (mappedTargetColumns.contains(mapping.getTargetColumn())) {
```

Comment:
[minor] You can call mappedTargetColumns.add(columnName) instead as it returns false when the entry already exists.

I confused this line with https://github.com/scalar-labs/scalardb/pull/2445/files#diff-fe051a3fcbc578fb1d270a16e032dbb16af0284fbd34ef077783848fba783138R89...